### DISCUSSION TITLE
Counting objects

### DISCUSSION THREAD

**Discussion by creativename00** [ultralytics/ultralytics#15016] [Q&A]
I'm working on a model to detect cracks in concrete box culverts. I have tried to make the different object counting methods work, but the environment doesn't really lend itself to it as the camera is what's moving and it's close to the object. Is there a way to just show the amount of detections at the end in a summary or similar. I don't really need the live view version.

**Comment by glenn-jocher** [comment]
@creativename00 you can achieve a summary of detections without a live view by processing the video frames and counting the detections at the end. Use the `model.predict` method to run inference on each frame, accumulate the counts, and then display the total count after processing all frames. This approach will allow you to focus on the final detection count without needing a real-time display. If you encounter any issues, ensure you are using the latest version of the Ultralytics package. For more detailed guidance, refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/).

**Reply by creativename00** [reply]
> @creativename00 you can achieve a summary of detections without a live view by processing the video frames and counting the detections at the end. Use the `model.predict` method to run inference on each frame, accumulate the counts, and then display the total count after processing all frames. This approach will allow you to focus on the final detection count without needing a real-time display. If you encounter any issues, ensure you are using the latest version of the Ultralytics package. For more detailed guidance, refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/).

Can you elaborate on this a bit? Obviously with just a frame it's easy enough to see how many detections of each type, but with a video that's where I'm struggling. Is there some coding that will accumulate and display the counts for all frames? This is my current code for running detections or segmentation.

`from ultralytics import YOLO

# Load a pretrained YOLOv8n model
model = YOLO('D:/bestsegment3-1.pt')

# Run inference on '____' with arguments
model.predict('D:\crack stock.mp4', save=True, save_txt=True, show=False, imgsz=640, conf=0.65)`

**Reply by glenn-jocher** [reply]
@creativename00 certainly! To accumulate and display the counts for all frames in a video, you can modify your code to iterate through each frame, run the detection, and keep a running total of the detections. After processing all frames, you can then print the total counts. Here's a minimal example to get you started:

```python
from ultralytics import YOLO
import cv2

# Load a pretrained YOLOv8n model
model = YOLO('D:/bestsegment3-1.pt')

# Open the video file
cap = cv2.VideoCapture('D:/crack stock.mp4')
assert cap.isOpened(), "Error reading video file"

total_detections = 0

while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break
    results = model.predict(frame, show=False, imgsz=640, conf=0.65)
    total_detections += len(results[0].boxes)

cap.release()
print(f"Total detections: {total_detections}")
```

This script processes each frame of the video, counts the detections, and accumulates the total count. Adjust the code as needed for your specific use case. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. For more detailed guidance, refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/).

**Reply by creativename00** [reply]
Thank you so much for that. Can you tell me what I would need to do in order to have it display the results by class? Can a tracker be integrated into this so it wouldn't recount the same crack?

**Reply by glenn-jocher** [reply]
You're welcome! To display results by class, you can modify the code to count detections per class. Integrating a tracker can help avoid recounting the same crack. Use the `model.track` method instead of `model.predict` to maintain object identities across frames. Ensure you are using the latest version of the Ultralytics package for the best results. If you need further assistance, the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/) provides detailed guidance.

**Reply by creativename00** [reply]
> You're welcome! To display results by class, you can modify the code to count detections per class. Integrating a tracker can help avoid recounting the same crack. Use the `model.track` method instead of `model.predict` to maintain object identities across frames. Ensure you are using the latest version of the Ultralytics package for the best results. If you need further assistance, the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/) provides detailed guidance.

I'm unfortunately still a bit lost on this one. I tried to merge the counting specific classes code provided in the documentation and the code you provided. What exactly needs to be changed in the one you provided to be able to show the count separated by class?

**Reply by glenn-jocher** [reply]
To count detections by class, you'll need to modify the code to keep track of counts for each class separately. Use the `model.track` method to maintain object identities across frames. Here's a minimal example to get you started:

```python
from collections import defaultdict
from ultralytics import YOLO
import cv2

# Load a pretrained YOLOv8n model
model = YOLO('D:/bestsegment3-1.pt')

# Open the video file
cap = cv2.VideoCapture('D:/crack stock.mp4')
assert cap.isOpened(), "Error reading video file"

class_counts = defaultdict(int)

while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break
    results = model.track(frame, show=False, imgsz=640, conf=0.65)
    for box in results[0].boxes:
        class_counts[box.cls] += 1

cap.release()
print(f"Class counts: {dict(class_counts)}")
```

This script processes each frame, tracks objects, and counts detections by class. Adjust as needed for your specific use case. For further details, refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/object-counting/).
